src/Codegen/DirectivesFinder.hack

```diff
@@ -0,0 +1,118 @@
+                // If the attribute is declared as `Bar/Baz`, then either:
+                //      1) Bar is a top-level namespace.
+                //      2) Bar is imported via a use namespace statement, e.g., `use namespace Foo\Bar`.
+                //
+                // In each case, we need to match the whatever comes before the first `\` in the attribute
+                // name (or the whole attribute name, if it contains no slashes) with whatever comes after
+                // the last `\` in each use type or use namespace statement.
+                //
+                // Once we have a match, we can build the fully qualified name by adding the attribute name
+                // to everything that comes before the last `\` in the eligible use statement.
+                //
+                $needle = Str\split($attribute, '\\')
+                    |> C\firstx($$);
+                $attribute = '\\'.$attribute;
+                foreach ($use_decls as $decl) {
+                    if (C\contains_key($decl['leaves'], $needle)) {
+                        // We have a match such as `Bar\Baz` with `use namespace Foo\Bar`;
+                        // build the qualified name as `Foo\Bar\Baz`.
+                        $attribute = '\\'.$decl['root'].$attribute;
+                        break;
+                    }
+                }
+            }
+
+            if (C\contains_key($this->directives, $attribute)) {
+                $directives[] = Str\format(
+                    'new %s(%s)',
+                    $attribute,
+                    Vec\map(
+                        $args,
+                        $arg ==> \var_export($arg, true)
+                            |> Str\replace($$, 'varray', 'vec')
+                            |> Str\replace($$, 'darray', 'dict')
+                            |> Str\replace($$, 'array', 'shape')
+                            |> Str\replace($$, 'vec ', 'vec')
+                            |> Str\replace($$, 'dict ', 'dict')
+                            |> Str\replace($$, 'shape ', 'shape')
+                    )
+                        |> Str\join($$, ', '),
+                );
+            }
+        }
+
+        return $directives;
+    }
+
+    <<__Memoize>>
+    private static async function getUseDeclarations(string $file_path): Awaitable<vec<shape(
+        'root' => string,
+        'leaves' => keyset<string>,
+    )>> {
+        $ast = await \Facebook\HHAST\from_file_async(\Facebook\HHAST\File::fromPath($file_path));
+        return $ast->getDescendantsByType<\Facebook\HHAST\INamespaceUseDeclaration>()
+            |> Vec\map($$, $decl ==> {
+                $kind = $decl->getKind();
+                if ($kind is null || $kind->getText() === 'function') {
+                    return null;
+                }
+                $leaves = $decl->getClausesx()->toVec();
+                if ($decl is \Facebook\HHAST\NamespaceGroupUseDeclaration) {
```

Comment:
HackAst is really slow and changes rapidly along with the language so a lot of it is dangerous to use. Can you use ReflectionClass::getNamespaceName() for this?

Reply:
At FB we ban the use of HackAst for anything that's production-critical, including build systems.

Reply:
Agreed using HHAST is a bummer. I can play around with ReflectionClass more, but I have to admit I'm not very optimistic about it working.

If developers manually registered their custom directives, then we could use `Reflection{Class, Method}::getAttributeClass`, which actually does what we want. So maybe that's the interim approach until we support Facts: when creating a new directive, developers register it with the generator:

```hack
await GraphQL\Codegen\Generator::forPath(
    __DIR__.'/Fixtures',
    shape(
        'output_directory' => __DIR__.'/gen',
        'namespace' => 'Slack\GraphQL\Test\Generated',
        'custom_directives' => shape(
            'fields' => vec[Directives\HasRole::class],
            'objects' => vec[Directives\HasRole::class],
        ),
    ),
);
```

That's starting to feel like the cleanest way of doing this.

Reply:
Registering feels totally reasonable

Reply:
Ok, did that in the latest commit: https://github.com/mwildehahn/hack-graphql/pull/123/commits/355b1412fe3c4b358cc7a5d41d7b0c9f2aeffac8

I like where this is headed now. Will work on object directives and introspection next.